pkg/api/v1beta3/types.go

```diff
@@ -509,6 +509,33 @@ type EnvVar struct {
 	Name string `json:"name" description:"name of the environment variable; must be a C_IDENTIFIER"`
 	// Optional: defaults to "".
 	Value string `json:"value,omitempty" description:"value of the environment variable; defaults to empty string"`
+	// Optional: specify a source for the value of the EnvVar.
+	*EnvVarSource `json:"source,omitempty" description:"source for the environment variable's value; cannot be used if value is set"`
+}
+
+// EnvVarSource represents a source for the value of an EnvVar.
+// Only one of its members may be specified.
+type EnvVarSource struct {
+	// ObjectInfo represents a downward API request for a field of the object.
+	//
+	// Valid fields of this reference are:
+	//
+	// 1. APIVersion
+	// 2. FieldPath
+	//
+	// No other fields may be specified.
+	//
+	// Currently ObjectInfo may be used to address the Name and Namespace
+	// only.
+	ObjectInfo *ObjectReference `json:"objectInfo,omitempty" description:"describes a field of the pod to use the value of; only APIVersion and FieldPath are accepted fields"`
```

Comment:
I like the idea of using ObjectReference.

If fields from the ObjectReference are omitted, they should default to fields matching the pod.

Why is APIVersion necessary?

Right now, it looks like FieldPath is half-baked and only defined for one field:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/container/ref.go#L47

We should make it match whatever syntax we settle on for kubectl #4348.